File: src/timerapp_ag/app_prefs.py
"""Локальные настройки приложения (не в data.json / WebDAV)."""
from __future__ import annotations

import json
import logging
from dataclasses import asdict, dataclass
from datetime import datetime, timezone
from pathlib import Path

from .platform_paths import config_dir
from .secure_files import write_json_secrets

logger = logging.getLogger(__name__)

MIN_UPDATE_CHECK_INTERVAL_DAYS = 1
MAX_UPDATE_CHECK_INTERVAL_DAYS = 30
DEFAULT_UPDATE_CHECK_INTERVAL_DAYS = 1
DEFAULT_UPDATE_GITHUB_REPO = "alexandrgert/timerapp_exp"


@dataclass
class AppPrefs:
    check_updates: bool = False
    update_check_interval_days: int = DEFAULT_UPDATE_CHECK_INTERVAL_DAYS
    update_github_repo: str = DEFAULT_UPDATE_GITHUB_REPO
    last_update_check_at: str = ""
    dismissed_update_version: str = ""

    def to_dict(self) -> dict:
        return asdict(self)

    @classmethod
    def from_dict(cls, data: dict | None) -> AppPrefs:
        payload = data if isinstance(data, dict) else {}
        interval = int(payload.get("update_check_interval_days") or DEFAULT_UPDATE_CHECK_INTERVAL_DAYS)
        interval = max(
            MIN_UPDATE_CHECK_INTERVAL_DAYS,
            min(MAX_UPDATE_CHECK_INTERVAL_DAYS, interval),
        )
        return cls(
            check_updates=bool(payload.get("check_updates", False)),
            update_check_interval_days=interval,
            update_github_repo=normalize_github_repo(
                str(payload.get("update_github_repo") or DEFAULT_UPDATE_GITHUB_REPO)
            ),
            last_update_check_at=str(payload.get("last_update_check_at") or ""),
            dismissed_update_version=str(payload.get("dismissed_update_version") or ""),
        )


def normalize_github_repo(value: str) -> str:
    text = (value or "").strip().removeprefix("https://github.com/").removeprefix("http://github.com/")
    text = text.strip("/")
    if text.count("/") != 1:
        return DEFAULT_UPDATE_GITHUB_REPO
    owner, name = text.split("/", 1)
    owner = owner.strip()
    name = name.strip().removesuffix(".git")
    if not owner or not name:
        return DEFAULT_UPDATE_GITHUB_REPO
    if not all(part.replace("-", "").replace("_", "").replace(".", "").isalnum() for part in (owner, name)):
        return DEFAULT_UPDATE_GITHUB_REPO
    return f"{owner}/{name}"


def app_prefs_path() -> Path:
    return config_dir() / "app.json"


def load_app_prefs() -> AppPrefs:
    path = app_prefs_path()
    if not path.is_file():
        return AppPrefs()
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        logger.warning("Failed to load app prefs: %s", exc)
        return AppPrefs()
    return AppPrefs.from_dict(payload if isinstance(payload, dict) else {})


def save_app_prefs(prefs: AppPrefs) -> None:
    write_json_secrets(app_prefs_path(), prefs.to_dict())


def parse_iso_datetime(value: str) -> datetime | None:
    text = (value or "").strip()
    if not text:
        return None
    try:
        if text.endswith("Z"):
            text = text[:-1] + "+00:00"
        return datetime.fromisoformat(text)
    except ValueError:
        return None


def should_run_auto_update_check(
    prefs: AppPrefs,
    *,
    now: datetime | None = None,
) -> bool:
    if not prefs.check_updates:
        return False
    stamp = now or datetime.now(timezone.utc)
    last = parse_iso_datetime(prefs.last_update_check_at)
    if last is None:
        return True
    if last.tzinfo is None:
        last = last.replace(tzinfo=timezone.utc)
    if stamp.tzinfo is None:
        stamp = stamp.replace(tzinfo=timezone.utc)
    elapsed = stamp - last
    return elapsed.total_seconds() >= prefs.update_check_interval_days * 86400


def mark_update_check_done(
    prefs: AppPrefs,
    *,
    now: datetime | None = None,
    dismissed_version: str | None = None,
) -> AppPrefs:
    stamp = now or datetime.now(timezone.utc)
    updated = AppPrefs(
        check_updates=prefs.check_updates,
        update_check_interval_days=prefs.update_check_interval_days,
        update_github_repo=normalize_github_repo(prefs.update_github_repo),
        last_update_check_at=stamp.astimezone(timezone.utc).isoformat(),
        dismissed_update_version=(
            dismissed_version
            if dismissed_version is not None
            else prefs.dismissed_update_version
        ),
    )
    save_app_prefs(updated)
    return updated

File: android/app/src/main/java/com/timerapp/linkb24/data/AppPrefs.kt
package com.timerapp.linkb24.data

import android.content.Context
import kotlinx.serialization.Serializable
import java.io.File
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.time.Instant
import java.time.OffsetDateTime
import java.time.format.DateTimeFormatter

const val MIN_UPDATE_CHECK_INTERVAL_DAYS = 1
const val MAX_UPDATE_CHECK_INTERVAL_DAYS = 30
const val DEFAULT_UPDATE_CHECK_INTERVAL_DAYS = 1
const val DEFAULT_UPDATE_GITHUB_REPO = "alexandrgert/timerapp_exp"

@Serializable
data class AppPrefs(
    val checkUpdates: Boolean = false,
    val updateCheckIntervalDays: Int = DEFAULT_UPDATE_CHECK_INTERVAL_DAYS,
    val updateGithubRepo: String = DEFAULT_UPDATE_GITHUB_REPO,
    val lastUpdateCheckAt: String = "",
    val dismissedUpdateVersion: String = "",
)

class AppPrefsRepository(
    private val prefsFile: File,
) {
    constructor(context: Context) : this(File(context.filesDir, "app.json"))

    fun load(): AppPrefs {
        if (!prefsFile.isFile) {
            return AppPrefs()
        }
        return runCatching {
            AppJson.decodeFromString(AppPrefs.serializer(), prefsFile.readText())
        }.getOrElse { AppPrefs() }.normalized()
    }

    fun save(prefs: AppPrefs) {
        prefsFile.parentFile?.mkdirs()
        val payload = AppJson.encodeToString(AppPrefs.serializer(), prefs.normalized())
        val tempFile = File(prefsFile.parentFile, "${prefsFile.name}.tmp")
        tempFile.writeText(payload)
        Files.move(
            tempFile.toPath(),
            prefsFile.toPath(),
            StandardCopyOption.REPLACE_EXISTING,
            StandardCopyOption.ATOMIC_MOVE,
        )
    }

    fun shouldRunAutoCheck(now: Instant = Instant.now()): Boolean {
        val prefs = load()
        if (!prefs.checkUpdates) {
            return false
        }
        if (prefs.lastUpdateCheckAt.isBlank()) {
            return true
        }
        val last = runCatching { OffsetDateTime.parse(prefs.lastUpdateCheckAt).toInstant() }
            .getOrNull() ?: return true
        val elapsedSeconds = now.epochSecond - last.epochSecond
        return elapsedSeconds >= prefs.updateCheckIntervalDays * 86400L
    }

    fun markCheckDone(
        dismissedVersion: String? = null,
        now: Instant = Instant.now(),
    ): AppPrefs {
        val current = load()
        val updated = current.copy(
            lastUpdateCheckAt = DateTimeFormatter.ISO_OFFSET_DATE_TIME.format(
                OffsetDateTime.ofInstant(now, java.time.ZoneOffset.UTC),
            ),
            dismissedUpdateVersion = dismissedVersion ?: current.dismissedUpdateVersion,
        )
        save(updated)
        return updated
    }
}

fun normalizeGithubRepo(value: String): String {
    var text = value.trim()
        .removePrefix("https://github.com/")
        .removePrefix("http://github.com/")
        .trim('/')
    if (text.count { it == '/' } != 1) {
        return DEFAULT_UPDATE_GITHUB_REPO
    }
    val parts = text.split('/', limit = 2)
    val owner = parts[0].trim()
    val name = parts[1].trim().removeSuffix(".git")
    if (owner.isEmpty() || name.isEmpty()) {
        return DEFAULT_UPDATE_GITHUB_REPO
    }
    fun ok(part: String): Boolean =
        part.replace("-", "").replace("_", "").replace(".", "").all { it.isLetterOrDigit() }
    if (!ok(owner) || !ok(name)) {
        return DEFAULT_UPDATE_GITHUB_REPO
    }
    return "$owner/$name"
}

fun AppPrefs.normalized(): AppPrefs {
    val days = updateCheckIntervalDays.coerceIn(
        MIN_UPDATE_CHECK_INTERVAL_DAYS,
        MAX_UPDATE_CHECK_INTERVAL_DAYS,
    )
    return copy(
        updateCheckIntervalDays = days,
        updateGithubRepo = DEFAULT_UPDATE_GITHUB_REPO,
    )
}
